gpt_engineer/ai.py

```diff
@@ -23,72 +39,134 @@ class TokenUsage:
 
-    def next(self, messages: List[Dict[str, str]], prompt=None, *, step_name=None):
+    def combine_messages(self, messages: List[Message]) -> str:
+        msg_dict = messages_to_dict(messages)
+        logging.debug(msg_dict)
+        prompt = "\n".join(
+            "%s: %s" % (md["type"], md["data"]["content"]) for md in msg_dict
+        )
+        logging.debug("Prompt: " + prompt)
+        return prompt
+
+    def next(
+        self,
+        messages: List[Message],
+        prompt: Optional[str] = None,
+        *,
+        step_name: str,
+    ) -> List[Message]:
         if prompt:
-            messages += [{"role": "user", "content": prompt}]
+            messages.append(self.fuser(prompt))
 
         logger.debug(f"Creating a new chat completion: {messages}")
-        response = openai.ChatCompletion.create(
-            messages=messages,
-            stream=True,
-            model=self.model,
-            temperature=self.temperature,
-        )
 
-        chat = []
-        for chunk in response:
-            delta = chunk["choices"][0]["delta"]  # type: ignore
-            msg = delta.get("content", "")
-            print(msg, end="")
-            chat.append(msg)
-        print()
-        messages += [{"role": "assistant", "content": "".join(chat)}]
+        msgs_as_prompt = self.combine_messages(messages)
+        response = self.llm(msgs_as_prompt, callbacks=[StreamingStdOutCallbackHandler()])
+        response = cleanup_reponse(response)
+
+        messages.append(self.fassistant(response))
+
         logger.debug(f"Chat completion finished: {messages}")
 
         self.update_token_usage_log(
-            messages=messages, answer="".join(chat), step_name=step_name
+            messages=messages, answer=response, step_name=step_name
         )
 
         return messages
 
-    def update_token_usage_log(self, messages, answer, step_name):
+    def last_message_content(self, messages: List[Message]) -> str:
+        m = messages[-1].content.strip()
+        logging.info(m)
+        return m
+
+    @staticmethod
+    def serialize_messages(messages: List[Message]) -> str:
+        return json.dumps(messages_to_dict(messages))
+
+    @staticmethod
+    def deserialize_messages(jsondictstr: str) -> List[Message]:
+        return AI.parse_langchain_basemessages(
+            messages_from_dict(json.loads(jsondictstr))
+        )
+
+    @staticmethod
+    def parse_langchain_basemessages(messages: Sequence[BaseMessage]) -> List[Message]:
+        """LangChain saves message history as Sequence[BaseMessage], which is immutable.
+        To make our code cleaner, we parse it to List[Message], which is mutable."""
+        parsed_messages = []
+
+        msg_type_to_cls = {
+            msg_cls(content="").type: msg_cls
+            for msg_cls in [AIMessage, HumanMessage, SystemMessage]
+        }
+        msg_types = list(msg_type_to_cls.keys())
+
+        for m in messages:
+            if m.type not in msg_types:
```

Comment:
Don't see this being necessary. We can assume it will work.

Reply:
Agreed & deleted